grammar/Expr.g4

```diff
@@ -17,23 +17,55 @@ grammar Expr;
 /* To match the whole input */
 fullexpr: expr EOF;
 
-shift: TIME (op=('+' | '-') expr)?;
-
-expr: '-' expr                  # negation
-    | expr op=('/' | '*') expr  # muldiv
-    | expr op=('+' | '-') expr  # addsub
-    | expr COMPARISON expr      # comparison
-    | IDENTIFIER                # identifier
-    | IDENTIFIER '.' IDENTIFIER # portField
-    | NUMBER                    # number
-    | '(' expr ')'              # expression
-    | IDENTIFIER '(' expr ')'   # function
-    | IDENTIFIER '[' shift (',' shift)* ']'  # timeShift
-    | IDENTIFIER '[' expr  (',' expr )* ']'  # timeIndex
-    | IDENTIFIER '[' shift1=shift '..' shift2=shift ']'      # timeShiftRange
-    | IDENTIFIER '[' expr '..' expr ']'      # timeRange
+expr
+    : atom                                     # unsignedAtom
+    | IDENTIFIER '.' IDENTIFIER                # portField
+    | '-' expr                                 # negation
+    | '(' expr ')'                             # expression
+    | expr op=('/' | '*') expr                 # muldiv
+    | expr op=('+' | '-') expr                 # addsub
+    | expr COMPARISON expr                     # comparison
+    | IDENTIFIER '(' expr ')'                  # function
+    | IDENTIFIER '[' shift (',' shift)* ']'    # timeShift
+    | IDENTIFIER '[' expr  (',' expr )* ']'    # timeIndex
+    | IDENTIFIER '[' shift1=shift '..' shift2=shift ']'     # timeShiftRange
+    | IDENTIFIER '[' expr '..' expr ']'        # timeRange
     ;
 
+atom
+    : NUMBER                                   # number
+    | IDENTIFIER                               # identifier
+    ;
+
+// a shift is required to be either "t" or "t + ..." or "t - ..."
+// Note: simply defining it as "shift: TIME ('+' | '-') expr" won't work
+//       because the minus sign will not have the expected precedence:
+//       "t - d + 1" would be equivalent to "t - (d + 1)"
+shift: TIME shift_expr?;
+
+// Because the shift MUST start with + or -, we need
+// to differentiate it from generic "expr".
+// A shift expression can only be extended to the right by a
+// "right_expr" which cannot start with a + or -,
+// unlike shift_expr itself.
+// TODO: the grammar is still a little weird, because we
```

Comment:
I left this TODO for further thought:
the grammar is not completely well balanced because we allow more stuff in the right parts than in the left parts.